### PR TITLE
Add new Environment Variable Event Handler and associated tests.

### PR DESCRIPTION
**Pull request by jonathannewell**
Moved code to an event handler as requested.

I wasn't sure on the package structure was desirable but chose to put the handler and its associated functions/tests under the event manager in a sub-dir structure.

Please advise if I need to refactor anything or add more tests. Test coverage is around 86%.

### CODE REVIEW

**Comment by jonathannewell** [comment]
@trayo I am not sure how this code coverage tool is calculating coverage. I have run the tests against my branch and the develop branch and in every area the coverage has remained the same or increased. The new stuff I added has 87.5% coverage which I believe meets the goal. Can you tell me what I need to do to get this to pass your checks? Cheers, Jon

**Comment by trayo** [comment]
@jonathannewell Don't worry about the code coverage. The script is a little wacky and doesn't always work right.

Is it possible for you to remove the extra features that were added. The title of the PR is "Add new Environment Variable Event Handler and associated tests." but I see some code that does some cleanup that is unrelated. It would be helpful to make the PR as small as possible so that it's easier for us to digest.

Your other tests didn't work, by the way, because they are more like integration tests which we haven't set up yet. We have an open source concourse pipeline that we want to use soon and that should help.

**Comment by jonathannewell** [comment]
@trayo The cleanup code is removed.

File: controller/deployer/deployer.go [hunk 1]
```diff
@@ -130,6 +131,13 @@ func (d Deployer) Deploy(req *http.Request, environment, org, space, appName, co
 		deploymentInfo.Instances = environments[environment].Instances
 	}
 
+	if deploymentInfo.HasEnvironmentVariables() {
```
Comment (trayo) [inline controller/deployer/deployer.go]: Is it possible to utilize the `deploy.start` event instead of creating an extra one? I'd like to see the new envvar handlers be decoupled from deployments.

File: controller/deployer/deployer.go [hunk 2]
```diff
@@ -151,10 +159,10 @@ func (d Deployer) Deploy(req *http.Request, environment, org, space, appName, co
 	defer emitDeployFinish(d, deployEventData, response, &err, &statusCode)
 
 	d.Log.Debug("emitting a deploy.start event")
-	err = d.EventManager.Emit(S.Event{Type: "deploy.start", Data: deployEventData})
+	err = d.EventManager.Emit(S.Event{Type: I.DEPLOY_START_EVENT, Data: deployEventData})
```
Comment (trayo) [inline controller/deployer/deployer.go]: I like the idea of these being constants! I think they should be defined at the top of deployer instead, since that is the only place they are used. Would you be able to send that as a separate pull request so that this pull request only adds the new handler?

File: controller/deployer/deployer.go [hunk 3]
```diff
@@ -203,21 +211,21 @@ func isJSON(contentType string) bool {
 }
 
 func emitDeployFinish(d Deployer, deployEventData S.DeployEventData, response io.ReadWriter, err *error, statusCode *int) {
-	d.Log.Debug("emitting a deploy.finish event")
+	d.Log.Debugf("emitting a %s event", I.DEPLOY_FINISH_EVENT)
```
Comment (trayo) [inline controller/deployer/deployer.go]: This change could also go into a PR for defining constants at the top of deployer.go

File: eventmanager/eventmanager.go [hunk 4]
```diff
@@ -26,6 +27,7 @@ func (e *EventManager) AddHandler(handler I.Handler, eventType string) error {
 		return InvalidArgumentError{}
 	}
 	e.handlers[eventType] = append(e.handlers[eventType], handler)
+	e.Log.Debugf("Handler for [%s] events successfully added!", eventType)
```
Comment (trayo) [inline eventmanager/eventmanager.go]: I think the log here should be `event` instead of `events`. That way it would read:

> Handler for [deploy.start] event successfully added!

File: eventmanager/handlers/envvar/manifest.go [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+	_, err = manifest.UnMarshal()
+
+	if err != nil {
+		logger.Errorf("Error Occurred during manifest creation/unmarshal! Details: %+v", err)
+	}
+
+	return manifest, err
+}
+
+func ReadManifest(path string, logger I.Logger, filesystem *afero.Afero) (manifest *Manifest, err error) {
```
Comment (trayo) [inline eventmanager/handlers/envvar/manifest.go]: Should this function exist in the tests if that's the only place it is used?

File: interfaces/constants.go [hunk 6]
```diff
@@ -0,0 +1,11 @@
+package interfaces
+
+const (
+	ENV_VARS_FOUND_EVENT                                         = "environment.variables.found"
+	DEPLOY_START_EVENT                                           = "deploy.start"
+	DEPLOY_FINISH_EVENT                                          = "deploy.finish"
+	DEPLOY_SUCCESS_EVENT                                         = "deploy.success"
+	DEPLOY_FAILURE_EVENT                                         = "deploy.failure"
+	ENABLE_ENV_VAR_HANDLER_FLAG_ARG                              = "env"
+	ENABLE_DISABLE_FILESYSTEM_CLEANUP_ON_DEPLOY_FAILURE_FLAG_ARG = "cleanOnFail"
```
Comment (trayo) [inline interfaces/constants.go]: I don't think `ENABLE_DISABLE_FILESYSTEM_CLEANUP_ON_DEPLOY_FAILURE_FLAG_ARG` is needed?

File: interfaces/constants.go [hunk 7]
```diff
@@ -0,0 +1,11 @@
+package interfaces
+
+const (
```
Comment (trayo) [inline interfaces/constants.go]: Same as the comment in `deployer.go` that I think these belong in the `deployer.go` file.

File: controller/deployer/deployer.go (hunk 1 above)
Comment (jonathannewell) [inline controller/deployer/deployer.go]: I think that will work. Curious, have you given any thought to having some sort of handler precedence mechanism if/when you get multiple handlers registered for the same event?

Comment (trayo) [inline controller/deployer/deployer.go]: The event manager will roll through all the handlers that are registered for an event. The precedence comes from where in the deployment you want your event to happen whether it is before, after, on failure or on success. (We might add another soon for in the middle of a deployment).

If you add a registered handler and it runs on `deploy.start` we treat and error returned as an event failure that's required before a deployment starts. If your event handler isn't necessary for a deployment and you don't want to fail a deployment when it runs, then the handler should always return `nil`.

You could create a precedence in `main.go` by reordering which handlers are registered first, since we just loop through an array of them.

File: controller/deployer/deployer.go (hunk 2 above)
Comment (jonathannewell) [inline controller/deployer/deployer.go]: The only problem I foresee with that is if we move them out of where they are you now have to have an import and reference to your Deployer in Server.go to register the handler..And Deployer imports lots of things....I tried something like that originally and ended up with circular references. I will give it a go though.

Comment (jonathannewell) [inline controller/deployer/deployer.go]: That worked. Will gen  a separate pull request for those shortly.

File: controller/deployer/deployer.go (hunk 3 above)
Comment (jonathannewell) [inline controller/deployer/deployer.go]: Will be done.

File: eventmanager/eventmanager.go (hunk 4 above)
Comment (jonathannewell) [inline eventmanager/eventmanager.go]: Done

File: interfaces/constants.go (hunk 7 above)
Comment (jonathannewell) [inline interfaces/constants.go]: Will remove this completely.

File: interfaces/constants.go (hunk 6 above)
Comment (jonathannewell) [inline interfaces/constants.go]: No longer required. Missed this when I removed the cleanup code.

File: controller/deployer/deployer.go (hunk 2 above)
Comment (trayo) [inline controller/deployer/deployer.go]: Oh! That makes sense actually. When I was looking for where we were using those, I only found them in `deployer.go` because we don't have event handlers in `main.go`. If you end up with circular problems then maybe they would be better suited on the event package?

I actually wouldn't mind them being in something similar to `interfaces`, it just seemed like a weird place for them. What if we made a `constants` package instead?

**Comment by trayo** [comment]
Regarding the Codecov again, I looked into and we added a `codecov.yml` but haven't made a new release on master yet so it will be a bit weird until then.

We plan on making a new release after this PR 😄

File: eventmanager/handlers/envvar/manifest.go (hunk 5 above)
Comment (jonathannewell) [inline eventmanager/handlers/envvar/manifest.go]: @trayo I just noticed this. Do you want me to move it?

**Comment by jonathannewell** [comment]
@trayo Ok, I have merge the upstream develop branch and made the changes your requested. Please advise if there is anything else I need to do. Thanks, Jon

File: structs/deployment_info.go [hunk 8]
```diff
@@ -3,19 +3,25 @@ package structs
...
+	Instances            uint16
+	Domain               string
+	AppPath              string
+	EnvironmentVariables map[string]string `json:"environment_variables"`
 
 	// Generic map used for users to provide their own deployment properties in JSON format.
 	Data map[string]interface{} `json:"data"`
 }
+
+func (info *DeploymentInfo) HasEnvironmentVariables() (bool) {
```
Comment (jonathannewell) [inline structs/deployment_info.go]: Yeah, that was instinctual based on many, many years of OOP. I will move it. Thanks!

**Comment by jonathannewell** [comment]
@trayo I know...I am high maintenance! Any word on this? Also, wanted to ask if your team/Deployadactyl committers were opposed to me creating a cmd line executable version of Deployadactyl and/or creating a capability for Event handlers to be out of process so that they can be developed separately? I am more than happy to have a call and discuss. Let me know. Cheers, Jon

**Comment by trayo** [comment]
@jonathannewell I'd be interested in what the executable would do. We had written one at one point and it merely extracted the `curl` to Go code and didn't end up being any less confusing. If we ever have to do a curl on the command line we use Postman to build the request and usually send it there. Otherwise, it has the ability to extract your request to code in many different languages (including a curl).

**Comment by jonathannewell** [comment]
As far as event handlers I was just thinking a rest api that allows a handler to register/unregister itself. So, basically, I (an event handler) call into Deployadactyl and say I am interested in xyz event. I provide Deploydactyl a URL where it will call me when that event occurs. The event api would be the same it just would occur via rest and allow the handlers to be out of process. This would prevent a handler from potentially abending Deploydactyl and allow handlers to be coded completely separately. If folks choose to contribute a repo could be setup to allow folks to contribute there handlers but it wouldn't be required. The only change I would suggest to the API would be something that allows handlers to specify a precedence when they register...not just FIFO.

The cmd line util thing for me is really about concerns I have about concurrent deploys to different environments. I know you spawn a new shell env. for each CF cmd execution. Have you had any issues with concurrent deploys to different target environments? Or to the same env. but different target org(s) or space(s)? As I think about it I am not sure just setting up Deploydactyl to be cmd line executable would solve that problem if it exists. I guess I need to do some experimentation.

**Comment by jonathannewell** [comment]
One other consideration on handlers. Probably need specify by event what data will be provided to handlers or allow the handler to tell the event manager what data it needs to be provided when the event is published. I like the second options even though it is more complex. Just need to specify by even what data is available. Meaning, some events would have access to more data than others. The data piece is the only outlier in the out of process model.

**Comment by trayo** [comment]
@jonathannewell Those seem like great ideas, I think we could move toward one of those solutions to eventing. Can you open github issue so we can track this discussion? Feel free to copy your comments as the description

**Comment by trayo** [comment]
Regarding the command line tool, it will be okay because we make each pusher use it's own `CF_HOME` that's specific to the deployment.

Unless you are thinking of writing a command line tool that will do the concurrent pushes and not a CURL to the deployadactyl server?
